### PR TITLE
[WIP] Switching to clang

### PR DESCRIPTION
**Pull request by General-Beck**
_Draft_
[parity-ethereum](https://gitlab.parity.io/General-Beck/parity-ethereum/pipelines/73889)
[substrate](https://gitlab.parity.io/General-Beck/substrate/pipelines/74960)
[polkadot](https://gitlab.parity.io/General-Beck/polkadot/pipelines/74907)

### CODE REVIEW

File: dockerfiles/base-ci-linux/Dockerfile [hunk 1]
```diff
@@ -38,10 +38,11 @@ RUN set -eux; \
 		>> /etc/apt/sources.list.d/llvm.list; \
 	apt-get -y update; \
 	apt-get install -y --no-install-recommends \
-		g++ libssl-dev gcc clang-9 libclang-9-dev make \
+		libssl-dev clang-9 clang++-9 lld-9 libclang-9-dev make \
 		git pkg-config curl time rhash; \
-# set a link to clang-9
-	ln -s /usr/bin/clang-9 /usr/bin/clang; \
+		# set a link to clang-9 and clang++-9
```
Comment (TriplEight) [inline dockerfiles/base-ci-linux/Dockerfile]:
```suggestion
	# set a link to clang-9 and clang++-9
```

File: dockerfiles/base-ci-linux/Dockerfile [hunk 2]
```diff
@@ -38,10 +38,11 @@ RUN set -eux; \
...
+		ln -s /usr/bin/clang-9 /usr/bin/clang; \
```
Comment (TriplEight) [inline dockerfiles/base-ci-linux/Dockerfile]:
```suggestion
	ln -s /usr/bin/clang-9 /usr/bin/clang; \
```

File: dockerfiles/base-ci-linux/Dockerfile [hunk 3]
```diff
@@ -38,10 +38,11 @@ RUN set -eux; \
...
+		ln -s /usr/bin/clang++-9 /usr/bin/clang++; \
```
Comment (TriplEight) [inline dockerfiles/base-ci-linux/Dockerfile]:
```suggestion
	ln -s /usr/bin/clang++-9 /usr/bin/clang++; \
```

File: dockerfiles/parity-ci-linux/Dockerfile [hunk 4]
```diff
@@ -18,11 +18,19 @@ dockerfiles/parity-ci-linux/README.md" \
...
+# rustup directory and compiler ENV
+#RUSTFLAGS Linker-plugin-LTO https://doc.rust-lang.org/rustc/linker-plugin-lto.html
+#RUSTDOCFLAGS issue https://github.com/rust-lang/cargo/issues/7529
 ENV RUSTUP_HOME=/usr/local/rustup \
     CARGO_HOME=/usr/local/cargo \
-    PATH=/usr/local/cargo/bin:$PATH
+    PATH=/usr/local/cargo/bin:$PATH \
+		RUSTFLAGS="-C linker-plugin-lto -C linker=clang -C link-arg=-fuse-ld=lld" \
+		RUSTDOCFLAGS="-C linker-plugin-lto -C linker=clang -C link-arg=-fuse-ld=lld" \
+    CC=clang \
```
Comment (TriplEight) [inline dockerfiles/parity-ci-linux/Dockerfile]:
```suggestion
		CC=clang \
```

**Comment by ddorgan** [comment]
@General-Beck are those android images in use by fether or the signer?

File: dockerfiles/ink-ci-linux/Dockerfile [hunk 5]
```diff
@@ -20,12 +20,12 @@ WORKDIR /builds
 
 # install specific Rust nightly toolchain
 # see on https://rust-lang.github.io/rustup-components-history/
+
```
Comment (TriplEight) [inline dockerfiles/ink-ci-linux/Dockerfile]: the comment above is related to the line below

File: dockerfiles/parity-ci-linux/Dockerfile [hunk 6]
```diff
@@ -18,11 +18,19 @@ dockerfiles/parity-ci-linux/README.md" \
...
+# config for clang
+COPY utility/parity-ci-linux-config /.cargo/config
+# rustup directory and compiler ENV
+#RUSTFLAGS Linker-plugin-LTO https://doc.rust-lang.org/rustc/linker-plugin-lto.html
+#RUSTDOCFLAGS issue https://github.com/rust-lang/cargo/issues/7529
 ENV RUSTUP_HOME=/usr/local/rustup \
     CARGO_HOME=/usr/local/cargo \
-    PATH=/usr/local/cargo/bin:$PATH
+    PATH=/usr/local/cargo/bin:$PATH \
+		RUSTFLAGS="-C linker-plugin-lto -C linker=clang -C link-arg=-fuse-ld=lld" \
```
Comment (TriplEight) [inline dockerfiles/parity-ci-linux/Dockerfile]:
```suggestion
	RUSTFLAGS="-C linker-plugin-lto -C linker=clang -C link-arg=-fuse-ld=lld" \
```

File: dockerfiles/parity-ci-linux/Dockerfile [hunk 7]
```diff
@@ -18,11 +18,19 @@ dockerfiles/parity-ci-linux/README.md" \
...
+		RUSTDOCFLAGS="-C linker-plugin-lto -C linker=clang -C link-arg=-fuse-ld=lld" \
```
Comment (TriplEight) [inline dockerfiles/parity-ci-linux/Dockerfile]:
```suggestion
	RUSTDOCFLAGS="-C linker-plugin-lto -C linker=clang -C link-arg=-fuse-ld=lld" \
```

File: dockerfiles/ink-ci-linux/Dockerfile (hunk 5 above)
Comment (TriplEight) [inline dockerfiles/ink-ci-linux/Dockerfile]:
```suggestion
```

**Pull request merged**